### PR TITLE
Add TimestampType and packing/unpacking support

### PR DESCRIPTION
**Pull request by randallpittman**
Here's my attempt at addressing #316. I made a TimestampType that can be packed/unpacked. I did not add any `datetime` conversions, but it will accept a floating-point seconds value for creation and can output to float with `to_float_s`. Also included a test. 

I tried to think it through pretty well and keep it 2/3 compatible. Let me know what you think.

### CODE REVIEW

File: msgpack/_packer.pyx [hunk 1]
```diff
@@ -253,6 +254,15 @@ cdef class Packer(object):
                     raise ValueError("EXT data is too large")
                 ret = msgpack_pack_ext(&self.pk, longval, L)
                 ret = msgpack_pack_raw_body(&self.pk, rawval, L)
+            elif type(o) is TimestampType if strict_types else isinstance(o, TimestampType):
+                # This should also be before Tuple because TimestampType is namedtuple.
+                code = -1
+                data = o.to_bytes()
```
Comment (jfolz) [inline msgpack/_packer.pyx]: This calls lots of Python functions, so it's probably quite slow.

**Comment by randallpittman** [comment]
Ok, sure, I wondered about that. I'll try to get that done in Cython.
Randy
________________________________

File: msgpack/__init__.py [hunk 2]
```diff
@@ -3,7 +3,12 @@
 from .exceptions import *
 
 from collections import namedtuple
+from numbers import Number, Integral
```
Comment (methane) [inline msgpack/__init__.py]: Don't use them.  Just use int.

File: msgpack/__init__.py [hunk 3]
```diff
@@ -3,7 +3,12 @@
...
 
+import struct
+import sys
+if sys.version_info[0] > 2:
+    long = int
```
Comment (methane) [inline msgpack/__init__.py]: I will drop Python 2 support before adding timestamp support.
So no need to care about Python 2 compatibility.

File: msgpack/__init__.py [hunk 4]
```diff
@@ -12,10 +17,74 @@ def __new__(cls, code, data):
             raise TypeError("code must be int")
         if not isinstance(data, bytes):
             raise TypeError("data must be bytes")
+        if code == -1:
+            return TimestampType.from_bytes(data)
         if not 0 <= code <= 127:
             raise ValueError("code must be 0~127")
         return super(ExtType, cls).__new__(cls, code, data)
 
+class TimestampType(namedtuple('TimestampType', 'seconds, nanoseconds')):
```
Comment (methane) [inline msgpack/__init__.py]: Don't use namedtuple.  Just use class with `__slots__`.

File: msgpack/_packer.pyx [hunk 5]
```diff
@@ -253,6 +254,15 @@ cdef class Packer(object):
                     raise ValueError("EXT data is too large")
                 ret = msgpack_pack_ext(&self.pk, longval, L)
                 ret = msgpack_pack_raw_body(&self.pk, rawval, L)
+            elif type(o) is TimestampType if strict_types else isinstance(o, TimestampType):
```
Comment (methane) [inline msgpack/_packer.pyx]: No need to `isinstance`.  strict_types is just a backward compatibility option.
No need to care about supporting subclass of Timestamp.

File: msgpack/__init__.py (hunk 2 above)
Comment (randallpittman) [reply msgpack/__init__.py]: May I ask why not? As a user I find it obnoxious to have to convert NumPy types to pure Python types and the `Number` and `Integral` general types alleviate that.

Comment (methane) [reply msgpack/__init__.py]: > May I ask why not? 

### Consistency

We don't use them in normal integers.  Why do we support it in Timestamp?
It doesn't make sense at all.

### Performance

```
$ python3 -m timeit -s 'from numbers import Integral' -- 'isinstance(42, Integral)'
500000 loops, best of 5: 443 nsec per loop
$ python3 -m timeit 'isinstance(42, int)'
5000000 loops, best of 5: 59.9 nsec per loop
```

---

>  As a user I find it obnoxious to have to convert NumPy types to pure Python types

Do you really want to use numpy types in Timestamp????

Comment (randallpittman) [reply msgpack/__init__.py]: Thanks, everything you said makes sense.

**Comment by randallpittman** [comment]
OK, I think I've addressed all the comments:
- I've added C packing and unpacking implementations (basically pulled from msgpack-c),
- Removed the Python 2 compat stuff (but still check for python3 in the test so it doesn't fail the CI.)
- TimestampType just expects basic python numeric types
- TimestampType is a simple class with `__slots__`.
- Added docs

Happy to get your thoughts.
-Randy

**Comment by methane** [comment]
Thank you, @randallpittman.  I merged this pull request with some modification.

**Comment by randallpittman** [comment]
Glad to contribute.

**Pull request closed without merging**